File: C++/cycledetect.cpp
#include <iostream>
#include <vector>
using namespace std;

class UnionFind {
private:
    vector<int> parent;  // Stores parent for each node

public:
    UnionFind(int n) : parent(n) {
        for (int i = 0; i < n; ++i) parent[i] = i;  // Each node is its own parent
    }

    int find(int x) {
        if (parent[x] != x)  // Path compression
            parent[x] = find(parent[x]);
        return parent[x];
    }

    bool unionSet(int x, int y) {
        int xr = find(x);  // Find root of x
        int yr = find(y);  // Find root of y
        if (xr == yr) return true;  // If same root, cycle exists
        parent[yr] = xr;  // Union the sets
        return false;
    }
};
int main() {
    int n = 5;
    UnionFind uf(n);

    vector<pair<int, int>> edges = {
        {0, 1}, {1, 2}, {2, 3}, {3, 4}, {2, 4}
    };

    for (auto& [u, v] : edges) {
        if (uf.unionSet(u, v)) {
            cout << "Cycle detected between " << u << " and " << v << endl;
        }
    }

    return 0;
}

File: Python/cycledetect.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.edges = []  # Store edges for later cycle detection

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX == rootY:
            return True  # Cycle detected
        self.parent[rootY] = rootX
        return False
    def add_edge(self, u, v):
        self.edges.append((u, v))

    def has_cycle(self):
        for u, v in self.edges:
            if self.union(u, v):
                return True
        return False



def main():
    print("Cycle Detection Demo")
    g = UnionFind(4)
    g.add_edge(0, 1)
    g.add_edge(1, 2)
    g.add_edge(2, 0)  # This creates a cycle

    print("Graph contains cycle?" , g.has_cycle())  # Should return True

if __name__ == "__main__":
    main()
